pkg/zng/docs/spec.md

```diff
@@ -1,92 +1,366 @@
+
+A "uvarint" is an unsigned integer encoded according to the protobuf specification.
+
+#### Record Type Encoding
 
-Any line beginning with `#` that does not conform with the syntax of a
-directive is an error.
-When errors are encountered parsing ZSON, an implementation should return a
-corresponding error and allow ZSON parsing to proceed if desired.
+Record types consist of an ordered set of columns where each column consists of
+a name and a typed value.  Unlike JSON, the ordering of the columns is significant
+and must be preserved through any APIs that consume, process, and emit ZNG records.
 
-## Regular Directives
+A record type is encoded as a count of fields followed by the field definitions,
+where a field definition is a field name followed by a type code.
 
-Regular directives have just three forms, each with a corresponding structure.
+The field names in a record must be unique.
 
-| Directive     | Example structure                   |
-|---------------|-------------------------------------|
-| Descriptor    | `#<int>:<type>`                     |
-| Ordering hint | `#sort [+-]<field>,[+-]<field>,...` |
-| Control       | `#!<payload>`                       |
+The count of fields is encoded as a uvarint.
 
-### Descriptor Directive
+The field name is encoded as a UTF8 string defining a "ZNG identifier"
+The UTF8 string
+is further encoded as a "counted string", which is uvarint encoding
+of the length of the string followed by that many bytes of UTF8 encoded
+string data.
 
-A descriptor directive defines the mapping between a decimal integer called
-a "descriptor" and a type according to the following form:
 ```
-#<descriptor>:<type>
+------------------------------------------------------------------
+|0x80|0x00|<num-fields>|<field1><typecode1><field2><typecode2>...|
+------------------------------------------------------------------
 ```
-There must be a single colon between the ASCII integer descriptor
-and the type definition, and the integer must be composed of a string of
-ASCII digits `0-9` with no leading `0` for any non-zero descriptor.
-The syntax for `<type>` is described by the [type grammar](#type-grammar).
-The same grammar applies to both regular types and legacy types (except for
-an exception regarding `.` in field names).
 
-The descriptor directive is the only directive that begins with an ASCII decimal
-digit.
+N.B.: The rules for ZNG identifiers follow the same rules as
+[JavaScript identifiers](https://tc39.es/ecma262/#prod-IdentifierName).
 
-For example, a directive that is a binding between descriptor `27`
-and a `record` composed of fields
-"foo" of type `string` and "bar" of type `int` is expressed as follows:
+The type code follows the field name and is encoded as a uvarint.
+
+A record may contain zero columns, in which case the only legal value
+for such a type is unset (see below).  An unset record with zero columns
+corresponds to a JSON empty object when translating ZNG to JSON and
```

Comment:
If we want to document ZJSON (i'd say we do) I think it would be clearer to have a standalone section for that, that comes later. Define bzng and zng semantics and syntax first, then show how that's represented in zjson. Right now it's a little confusing to see this json reference pop up here (and a couple other places).

Reply:
This part transcends zjson.  I'm talking about arbitrary JSON being translated into BZNG.  But, yes, this interpretative text can go elsewhere.

Reply:
+1